drivers/power/ltc4162l/ltc4162l.c

```diff
@@ -0,0 +1,829 @@
+		*val = voltage;
+
+		return 0;
+	case ID_LTC4162S:
+		voltage = 6000000 + (regval * 28571);
+		voltage *= cell_count/2;
+		*val = voltage;
+
+		return 0;
+	case ID_LTC4015:
+		if (chem_type <= 3)
+			voltage = 3812500 + (regval * 12500);
+		else if (chem_type > 3 || chem_type < 7)
+			voltage = 3412500 + (regval * 12500);
+		else
+			voltage = 2000000 + (regval * 9523);
+
+		voltage *= cell_count;
+		*val = voltage;
+		return 0;
+	default:
+		return -EINVAL;
+	}
+}
+
+/**
+ * @brief Set the vcharge setting
+ * @param dev - The LTC4162L device structure.
+ * @param val - The vcharge setting.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ltc4162l_set_vcharge_setting(struct ltc4162l_desc *dev, uint16_t val)
+{
+	uint32_t base_voltage, scale_factor, range;
+	uint16_t cell_count;
+	uint16_t chem_type;
+	int ret;
+
+	ret = ltc4162l_get_cell_count(dev, &cell_count);
+	if (ret)
+		return ret;
+
+	ret = ltc4162l_get_chem_type(dev, &chem_type);
+	if (ret)
+		return ret;
+
+	switch (dev->type) {
+	case ID_LTC4162L:
+		base_voltage = 3812500;
+		scale_factor = 12500;
+		range = 31;
+		break;
+	case ID_LTC4162F:
+		base_voltage = 3412500;
+		scale_factor = 12500;
+		range = 31;
+		break;
+	case ID_LTC4162S:
+		base_voltage = 6000000;
+		scale_factor = 28571;
+		range = 31;
+		break;
+	case ID_LTC4015:
+		if (chem_type <= 3) {
+			base_voltage = 3812500;
+			scale_factor = 12500;
+			range = 31;
+		} else if (chem_type < 7) {
+			base_voltage = 3412500;
+			scale_factor = 12500;
+			range = 31;
+		} else {
+			base_voltage = 2000000;
+			scale_factor = 9523;
+			range = 35;
+		}
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	cell_count = dev->type == ID_LTC4162S ? cell_count / 2 : cell_count;
+
+	if (!cell_count)
+		return -EBUSY; /* Not available yet, try again later */
+
+	/* Adjust value based on cell count and limits */
+	val /= cell_count;
+	if (val < base_voltage)
+		return -EINVAL;
+
+	val = (val - base_voltage) / scale_factor;
+	if (val > range)
+		return -EINVAL;
+
+	return ltc4162l_reg_write(dev, LTC4162L_VCHARGE_SETTING, val);
+}
+
+/**
+ * @brief Get the iin limit
+ * @param dev - The LTC4162L device structure.
+ * @param val - The iin limit.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ltc4162l_get_iin_limit(struct ltc4162l_desc *dev, uint16_t *val)
+{
+	uint16_t regval;
+	int ret;
+
+	ret = ltc4162l_reg_read(dev, LTC4162L_IIN_LIMIT_DAC, &regval);
+	if (ret)
+		return ret;
+
+	regval = no_os_field_get(LTC4162L_IIN_LIMIT_DAC_MASK, regval);
+
+	++regval;
+	regval *= 5000000u;
+	regval /= dev->rsnsi;
+	*val = 100u * regval;
+
+	return 0;
+}
+
+/**
+ * @brief Set the iin limit
+ * @param dev - The LTC4162L device structure.
+ * @param val - The iin limit.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ltc4162l_set_iin_limit(struct ltc4162l_desc *dev, uint16_t val)
+{
+	uint16_t regval;
+
+	regval = LTC4162L_MULT_FRAC(val, dev->rsnsi, 50000u);
+	regval /= 10000u;
```

Comment:
Put chip dependent constants in chip_info especially if these values are different between variants.